torch/csrc/jit/script/compiler.cpp

```diff
@@ -121,6 +142,60 @@ struct Environment {
     }
   }
 
+  void setBoolInfo(Value* v, BoolInfo info) {
+    bool_information[v] = std::move(info);
+  }
+
+  c10::optional<BoolInfo> getBoolInfo(Value* v) {
+    for (auto runner = this; runner; runner = runner->next.get()) {
+      auto info = runner->bool_information.find(v);
+      if (info != runner->bool_information.end()) {
+        return info->second;
+      }
+    }
+    return c10::nullopt;
+  }
+
+  void setRefinements() {
+    if (getBlockOwningKind() != prim::If)
+      return;
+
+    auto cond_value = b->owningNode()->input();
+
+    // only allow expressions within the conditional to refine types
+    // do not allow assigning expression to a variable and then using variable
+    if (cond_value->hasUniqueName())
+      return;
+
+    auto maybe_bool_info = getBoolInfo(cond_value);
+    if (maybe_bool_info) {
+      auto bool_info = *maybe_bool_info;
+      auto is_true_block = b->owningNode()->blocks().at(0) == b;
+      if (is_true_block) {
+        active_refinements = &maybe_bool_info->true_refinements;
+      } else {
+        active_refinements = &maybe_bool_info->false_refinements;
+      }
+      insertRefinements();
+    }
+  }
+
+  void insertRefinements() {
```

Comment:
This is a weird API. It is only called in one place. Why not have it take active_refinements as an argument?